### PR TITLE
Added RevokableTokenLock tests

### PR DESCRIPTION
**Pull request by STYJ**

### CODE REVIEW

File: test/RevokableTokenLock.spec.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    });
+    it('should not revert if caller is revoker', async () => {
+      await expect(revokableTokenLock.connect(revoker).revoke(recipient.address)).to.not.be.reverted;
+    });
+    it('should not revert if caller is owner', async () => {
+      await expect(revokableTokenLock.revoke(recipient.address)).to.not.be.reverted;
+    });
+    it('should claim any tokens vested and revoke the rest when revoke is called', async () => {
+      // TODO: how to move this into a fixture?
+      await token.approve(revokableTokenLock.address, amount);
```
Comment (MrToph) [inline test/RevokableTokenLock.spec.ts:83]: what's the problem with just adding it to the `fixture()` above?

File: test/RevokableTokenLock.spec.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        token,
+        [recipient, owner],
+        [transferred, remaining]
+      );
+
+      // Check contract values
+      ({unlockEnd, claimedAmounts, lockedAmounts} = await revokableTokenLock.vesting(recipient.address));
+      expect(claimedAmounts).to.equal(transferred);
+      expect(lockedAmounts).to.equal(claimedAmounts);
+      expect(unlockEnd).to.equal(unlockCliff);
```
Comment (MrToph) [inline test/RevokableTokenLock.spec.ts:98]: I think the test should check fewer implementation details and more functionality:
1) the claimer's balance increased by `amounts`
2) the owner's balance increased by the `remaining`
3) if you keep moving forward in time and check `claimableBalance`, it should return 0

File: test/RevokableTokenLock.spec.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      await revokableTokenLock.lock(recipient.address, amount);
+      await setNextBlockTimestamp(unlockCliff.toNumber());
+
+      // Call revoke
+      expect(await revokableTokenLock.revoke(recipient.address))
+        .to.emit(revokableTokenLock, 'Revoked')
+        .withArgs(recipient.address, remaining);
+    });
+    // TODO: mock revert on transfers to test transfer fails
+  });
```
Comment (MrToph) [inline test/RevokableTokenLock.spec.ts:112]: Can we also add another test that revoking _before_ the cliff returns zero and reclaims everything?

File: test/shared/Constants.ts [hunk 4]
```diff
@@ -7,3 +7,4 @@ export const ZERO = ethers.constants.Zero;
 export const ONE = ethers.constants.One;
 export const TWO = ethers.constants.Two;
 export const MAX_UINT = ethers.constants.MaxUint256;
+export const HOUR = ethers.BigNumber.from(60).mul(60).mul(1000);
```
Comment (MrToph) [inline test/shared/Constants.ts:10]: this is in milliseconds, see other comment

**Review by MrToph — COMMENTED** [review]
Thanks for working on this. Looks good mostly, just some suggestions for further tests

File: test/RevokableTokenLock.spec.ts (hunk 3 above)
Comment (MrToph) [reply test/RevokableTokenLock.spec.ts:112]: Maybe also one that tries to revoke after the entire vesting schedule is over, and check that it does not lead to any weird resets where the user could claim again.

File: test/RevokableTokenLock.spec.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    const revokableTokenLock = (await RevokableTokenLockFactory.deploy(
+      token.address,
+      revoker.address
+    )) as RevokableTokenLock;
+
+    const dt = new Date();
+    const begin = ethers.BigNumber.from(dt.getTime()).add(HOUR);
+    const cliff = begin.add(HOUR.mul(5));
+    const end = begin.add(HOUR.mul(20));
+    await revokableTokenLock.setupVesting(recipient.address, begin, cliff, end);
```
[the remark this answers is not in the document]
Comment (STYJ) [reply test/RevokableTokenLock.spec.ts:30]: Ah good catch, let me fix.

File: test/RevokableTokenLock.spec.ts (hunk 1 above)
Comment (STYJ) [reply test/RevokableTokenLock.spec.ts:83]: No big issue, just not used for other tests. I wanted to move this into its own fixture but am having issues trying to load 2 different snapshots. Will move into the earlier fixture.

File: test/RevokableTokenLock.spec.ts (hunk 2 above)
Comment (STYJ) [reply test/RevokableTokenLock.spec.ts:98]: 1 and 2 are actually checked on lines 88-92!

Will add test for 3.

File: test/RevokableTokenLock.spec.ts (hunk 3 above)
Comment (STYJ) [reply test/RevokableTokenLock.spec.ts:112]: done

File: test/shared/Constants.ts (hunk 4 above)
Comment (STYJ) [reply test/shared/Constants.ts:10]: noted

**Comment by STYJ** [comment]
Btw I'm having some troubles getting smock (https://github.com/defi-wonderland/smock) to work nicely w/ fixtures so this `TODO: mock revert on transfers to test transfer fails` is still a WIP.

I think this should be fine because the underlying token that is being claimed is ArenaToken and does not have any weird functionality like fee on transfers. Moreover, this scenario (having less token balance in the TokenLock contract than is tracked) can only happen if some user is able to claim more than intended which is not possible unless there is a logic flaw in the internal accounting.

**Comment by MrToph** [comment]
Agree that we don't need to test the transfer reverts

File: test/RevokableTokenLock.spec.ts [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    await token.approve(revokableTokenLock.address, amount);
+    await revokableTokenLock.lock(recipient.address, amount);
+
+    return {token, revokableTokenLock};
+  }
+  beforeEach('deploy fixture', async () => {
+    ({token, revokableTokenLock} = await loadFixture(fixture));
+  });
+  // TODO: maybe do getStorageAt() tests as well
```
Comment (MrToph) [inline test/RevokableTokenLock.spec.ts:41]: do we want to? what would these be used for? I prefer testing for functionality instead, not that important that a specific value at a specific storage slot is set

File: test/RevokableTokenLock.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (MrToph) [reply test/RevokableTokenLock.spec.ts:32]: All these time values also don't need to be `BigNumber`s, the easily fit into JS's integers and they are much easier to work with as you can just use the arithmetic operators `+`. Only the token amounts need to be

File: test/RevokableTokenLock.spec.ts [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        [amount, 0]
+      );
+    });
+    it('should return claimable balance of 0 after some time has passed since revoke was called.', async () => {
+      await setNextBlockTimeStamp(unlockCliff.toNumber());
+      await revokableTokenLock.revoke(recipient.address);
+
+      // Set next block timestamp to 2 hour after revoke was called and mine a block.
+      await setNextBlockTimeStamp(unlockCliff.add(HOUR.mul(TWO)).toNumber());
+      await mineBlocks(ONE.toNumber());
```
Comment (MrToph) [inline test/RevokableTokenLock.spec.ts:113]: There's a shared `mineAtBlock` function in the `TokenLock` test PR. If this is merged, might just want to use this one instead. https://github.com/code-423n4/contracts/pull/29/files#diff-6d7a55d4a3b08bd886254229dc52ae7929824c0c293304d75ae168864c43b0edR7

File: test/RevokableTokenLock.spec.ts [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+      const bal = await revokableTokenLock.claimableBalance(recipient.address);
+      expect(bal).to.equal(0);
+    });
+    it('should emit the Revoked event when revoke is called', async () => {
+      await setNextBlockTimeStamp(unlockCliff.toNumber());
+      expect(await revokableTokenLock.revoke(recipient.address))
+        .to.emit(revokableTokenLock, 'Revoked')
+        .withArgs(recipient.address, remaining);
+    });
+    // TODO: mock revert on transfers to test transfer fails
```
Comment (MrToph) [inline test/RevokableTokenLock.spec.ts:123]: can remove this TODO as we don't need to test for this imo

Comment (STYJ) [reply test/RevokableTokenLock.spec.ts:123]: done

File: test/RevokableTokenLock.spec.ts (hunk 6 above)
Comment (STYJ) [reply test/RevokableTokenLock.spec.ts:41]: getStorageAt() tests assumes that there's some error w/ how values are stored on the contract. This is definitely a nice to have but is non trivial to do and not as important now.

File: test/RevokableTokenLock.spec.ts (hunk 7 above)
Comment (STYJ) [reply test/RevokableTokenLock.spec.ts]: updated!
